Issue #301: Tap to delete media

aforcier opened the pull request:
Fixes #301, adding a 'delete' button to the overlay when an image is selected:

![301-tap-to-delete-media](https://cloud.githubusercontent.com/assets/9613966/15073530/f891a694-1367-11e6-875e-30dc3d4b7b91.png)

Tapping the delete button removes the image from the post.

<h4>Notes</h4>

- Does not work for video. Videos don't have a settings screen, and adding an overlay when tapping them is tricky since tapping plays/pauses the video. A solution I see for this is to permanently overlay a delete button on the top right of videos. I'll be opening a separate issue.
- I moved all the inline SVGs in the editor CSS (edit buttons) to standalone SVG files for better readability

cc @maxme and @mattmiklic for design feedback

aforcier committed:
Fixed issue where image settings screen wouldn't load in the example app

aforcier committed:
Added delete button to top right of images when they're selected (and the edit overlay is visible)

- The delete button is the 'cross-small' gridicon: http://automattic.github.io/gridicons/

aforcier committed:
Color image delete button red

aforcier committed:
Remove image from document when delete button is tapped

aforcier committed:
Adjusted media delete button handling to avoid deleting anything beyond the media's containing paragraph

aforcier committed:
Don't add delete overlay for very small images

- On very small images the delete overlay will take up most of the area, making it very hard to reach the image settings screen rather than delete the image

aforcier committed:
Clear currentEditingImage when an image is deleted

aforcier committed:
When deleting an image, run emptyFieldIfNoContents to clear any empty divs remaining, so the placeholder will re-appear

aforcier committed:
Rely on node.parent.removeChild(node) rather than node.remove() since the latter is unsupported on API<19

- Also cleaned up node traversal for media deletions for clarity

aforcier committed:
Replaced inline svgs in the CSS with files in the /svg/ directory

aforcier committed:
Moved wpposter.svg to /svg/ folder

aforcier committed:
Fixed alignment mistake in delete image button

maxme commented on libs/editor-common/assets/editor.css:
```diff
@@ -143,12 +143,12 @@ span.img_container.failed::before {
     pointer-events: none;
     content:"";
-    background:url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="54" height="54" viewBox="0 0 124 124"><circle fill="none" stroke="#FFFFFF" stroke-width="4"  cx="62" cy="62" r="60"/><path d="M79.0637899,40.93125 C74.1838649,36.0375 67.4465291,33 59.9831144,33 C45.0562852,33 33,45.0825 33,60 C33,74.9175 45.0562852,87 59.9831144,87 C72.5628518,87 83.0994371,78.39375 86.088 ...
+    background:url('svg/retry-image.svg') no-repeat center top;
```
+1

mattmiklic commented:
Tried a couple of ideas to help the "remove" icon better match the existing Edit icon, so they feel more related. I don't think red is necessary here since it's not destructive (no data is being deleted), the user can always re-add the photo to the post. 

Mockup:
![mockup](https://cloud.githubusercontent.com/assets/1396552/15087139/1519c26e-13b4-11e6-9a49-35367e76d2ff.png)

I added a 10% black background to the icons, which I think would make sure they're always visible, even on light images. 

Measurements:
<img width="525" alt="screen shot 2016-05-06 at 5 56 42 pm" src="https://cloud.githubusercontent.com/assets/1396552/15087142/209e5488-13b4-11e6-974e-71bc155463b4.png">

I've exported the `gridicons-cross` and `gridicons-pencil` icons at the needed sizes here: https://cloudup.com/ceuV4iy2USZ

aforcier commented:
That looks much better :grinning:.

One request: could I get the icons in `svg` format since that's how things are set up within the visual editor (they're loaded in the CSS)?

Also what are your thoughts on updating the retry button along with the others?

![187-media-upload-fail-overlay](https://cloud.githubusercontent.com/assets/9613966/9488169/97196c5c-4be0-11e5-8d86-0e15810089ed.png)

All the current assets (as of this branch) are here, for reference:

https://github.com/wordpress-mobile/WordPress-Editor-Android/tree/7b6c732360e8d5ab98cf07e133aaba92f2aaa2c2/libs/editor-common/assets/svg

There are two sizes for each of the retry and edit icons in there, for different sizes of images.

mattmiklic commented:
> One request: could I get the icons in svg format since that's how things are set up within the visual editor (they're loaded in the CSS)?

Absolutely. Actually, if you need SVGs we can just use the originals direct from Gridicons:

`gridicons-cross`:
https://github.com/Automattic/gridicons/blob/master/svg-min/gridicons-cross.svg

`gridicons-pencil`:
https://github.com/Automattic/gridicons/blob/master/svg-min/gridicons-pencil.svg

mattmiklic commented:
Just in case you need the SVGs already colored and resized I've re-exported them here, too. https://cloudup.com/cwFmnARhgOi

aforcier committed:
Updated delete image button

- Changed size, and changed color from red to white
- Added a circle around the X, generated using CSS
- Added 10% black background fill to improve visibility on lighter images

aforcier committed:
Updated edit image icon (normal size)

- Updated from dashicon to gridicon
- The circle around the pencil icon is now CSS-generated rather than built into the SVG
- Added 10% black background for better visibility on light images

aforcier committed:
Moved edit-icon css under edit-overlay

aforcier committed:
Use the new (gridicon) edit icon for small images too

- Since the default icon is now smaller, updated to use the same icon on small images, but dropping the circle border and the caption
- This makes the edit icon on small images slightly larger, 32x32 instead of 27x27

aforcier committed:
Adjusted font-weight for 'Edit' text to match Roboto Medium

aforcier commented:
![301-edit-overlay-white](https://cloud.githubusercontent.com/assets/9613966/15154891/ef85ad0e-16ad-11e6-95de-ade29884e2cb.png)

<h3>Done:</h3>
- [x] Changed the edit pencil icon from dashicon to gridicon
- [x] Switched to adding the circle border around edit icon in CSS rather than inside the SVG itself
- [x] Changed close button style to match @mattmiklic's specs
- [x] Added `10%` dark background to edit and delete icons
- [x] Adjusted the font of the `Edit` text to Medium (was set to weight 600, it's now 500)
- [x] Adjusted edit button for small images (less than `100dp` on any side), it's now the same as the normal sized edit button (but without a caption or the circle border)

<h3>Remaining:</h3>
- [x] Switch retry button from dashicon to gridicon
- [x] Add `10%` dark background to retry button
- [x] Update retry icon for large images accordingly
- [x] Fix backspace to delete images that have their edit overlay visible 

Re: the last point, I found that backspacing an image that's selected (so that the edit overlay is visible) leaves the container HTML in the document. This seems to be caused by adding the `delete-overlay` element to the `edit-container`, and the fix will probably require tracking the edit overlay with the `MutationObserver`.

cc @mattmiklic for another look

mattmiklic commented:
Looking good to me now. 👍

aforcier committed:
When selecting an image, add the image node to the MutationObserver to track manual deletion

- This fixes a bug with backspacing to delete selected images on API 19+ (the container wasn't being deleted)

aforcier committed:
Update retry icon from dashicon to gridicon

- Circle around icon is now CSS-based
- Added 10% dark background for better visibility

aforcier committed:
Adjusted spacing between retry icon and text

aforcier committed:
Adjusted 'Tap to try again!' text weight to Medium

aforcier committed:
Updated image upload retry button for large images to use a gridicon

aforcier committed:
Duplicated changes to image retry button for videos

aforcier committed:
Fixed missing 10% dark background for large retry buttons

aforcier commented:
Updated:

![372-retry-gridicon](https://cloud.githubusercontent.com/assets/9613966/15229354/abb4e90c-185e-11e6-995d-cc04e97dcfb4.png)
- [x] Retry button for failed media (normal and large) is now a gridicon, and I applied the same style changes as for the edit button
- [x] Fixed bug with manually deleting selected images

aforcier committed:
Clear currentEditingImage when launching the image settings dialog

- This fixes a bug where tapping the same image again after returning from image settings would show image settings again (instead of showing the edit image overlay)

aforcier commented:
Also fixed a bug where selecting an image > pressing edit > returning from the edit settings screen > tapping the same image would show the edit settings screen again, instead of showing the edit overlay.

maxme commented:
New remove button works well, but I think the Edit image settings screen doesn't save anymore, I'll check without the patch.

maxme commented:
This commit 69f55745a0c1522e7dfb266a1f9c43bc6453abc3 is the culprit, apparently it's too early to set `currentEditingImage` to null here, update image will be called later and image edits won't be saved.

aforcier committed:
When the image settings dialog is dismissed, clear the currentEditingImage in the ZSSEditor

- Corrects 69f557, which erroneously cleared the currentEditingImage before the image settings changes could be applied

aforcier commented:
> This commit 69f5574 is the culprit, apparently it's too early to set currentEditingImage to null here, update image will be called later and image edits won't be saved.

Fixed in https://github.com/wordpress-mobile/WordPress-Editor-Android/pull/372/commits/b250069b1d1be53e35629550e1a0eab7546d7d2e. You were right, it was too early, [`updateCurrentImageMeta()`](https://github.com/wordpress-mobile/WordPress-Editor-Android/blob/b250069b1d1be53e35629550e1a0eab7546d7d2e/libs/editor-common/assets/ZSSRichTextEditor.js#L1900) relies on `currentEditingImage` when the settings dialog is closed and changes are saved. `updateCurrentImageMeta()` clears `currentEditingImage` when it's done; the problem was that when the dialog is dismissed without saving the ZSSEditor isn't notified. That should work correctly now.

maxme commented:
:shipit:
